### PR TITLE
Add sample dashboard playwright test

### PR DESCRIPTION
**Pull request by adamint**
Runs dashboard with a mocked dashboard client, adds example playwright test, as we will be adding many more dashboard playwright tests using the same infrastructure in the future.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4386)

### CODE REVIEW

**Comment by Meir017** [comment]
related to #958

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs
Comment (Meir017) [inline tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:11]: https://playwright.dev/dotnet/docs/test-runners#xunit-support It might be better use one of the officially supported testing-libraries for playwright (MSTest or NUnit https://playwright.dev/dotnet/docs/test-runners#xunit-support)

Comment (adamint) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:11]: I would definitely prefer to use MSTest or NUnit, but I do want to balance that against other concerns that
- this would be the only test project not using xUnit, and
- there are other playwright tests using xUnit

@danmoseley Could we get your opinion on this?

Comment (Meir017) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:11]: @mxschmitt whats your opinion here?

Comment (radical) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:11]: > 
> > there are other playwright tests using xUnit
> 
> Is this referring to [that](https://github.com/dotnet/aspire/blob/521a2c63cbcc3dcbc5d93c99d8c6fbc3a0933b39/tests/Aspire.Workload.Tests/WorkloadTestsBase.cs#L26)?

Yes. And the Workload tests are using Playwright in (b) mode from your comment. We `PackageReference` `"Microsoft.Playwright"` nuget package, and use that.

Comment (adamint) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:11]: Yes, the concern is whether the benefits of switching to MSTest or nUnit outweigh the cost of supporting a singular test project not built with xUnit.

Comment (danmoseley) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:11]: > @danmoseley Could we get your opinion on this?

I don't have much to add: Aspire builds with reusable infrastructure from dotnet/arcade, which supports xunit, as the .NET platform repos picked xunit long ago. I do not see anything in there for mstest (as a test platform) or nunit so use of those would presumably have to be set up custom in this repo.

I defer to @radical/@joperezr

Comment (radical) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:11]: I'm not sure what additional work would be needed on the CI to support that - on the build machine, and helix. Also, arcade provides many test attributes for xunit like `ActiveIssue` which we use to skip tests conditionally. We also have `RequiresDocker` which is xunit based too.

This would also mean additional documentation, and providing ways to have similar functionality for skipping tests etc.

IMHO, sticking to xunit would be better as it would utilize all the existing infrastructure.

Comment (mxschmitt) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:11]: Sounds reasonable! As long as you provide a base class which gives them a browser context and test per test it should not be too bad of a compromise. Also make sure to add an expect method like [here](https://github.com/microsoft/playwright-dotnet/blob/86ad6c744350f97b01eba68dbb3b4c59efc9caf4/src/Playwright.MSTest/PlaywrightTest.cs#L129).

**Comment by adamint** [comment]
Could I get a review on this?

**Comment by radical** [comment]
Build failing with: `tests/Aspire.Dashboard.Tests/Integration/Playwright/DashboardServerFixture.cs(30,35): error CS0117: (NETCORE_ENGINEERING_TELEMETRY=Build) 'DashboardConfigNames' does not contain a definition for 'DashboardOtlpUrlName'`

**Comment by adamint** [comment]
> DashboardOtlpUrlName

Sorry, @radical. That property name was seemingly changed and I did not notice after merge. It builds now 😄

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs
Comment (radical) [inline tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:26]: On the linux (ubuntu) build machine `install-deps` will install a whole bunch of packages with `apt-get`. Because this is a shared machine, we want to avoid installing packages on it.

The existing infrastructure for installing playwright available is partially broken right now due to package installation issues (https://github.com/dotnet/aspire/issues/4623). And these tests are likely to run into the same issue randomly.

So, I would suggest not running the Playwright based tests on linux for now. You can add this to your test - `[ActiveIssue("https://github.com/dotnet/aspire/issues/4623", TestPlatforms.Linux)]`.

Going forward:
1. https://github.com/dotnet/aspire/pull/4689 - this will make the existing playwright targets reusable
2. Once https://github.com/dotnet/dnceng/issues/3122 is resolved, we can drop even this code.
Neither of these need to block your PR though.

Comment (radical) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:26]: Also, could you please open an issue to track that the code here needs to be adjusted once https://github.com/dotnet/aspire/pull/4689 is merged, and https://github.com/dotnet/dnceng/issues/3122 is fixed?

**Review by radical — DISMISSED** [review]

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/DashboardServerFixture.cs
Comment (adamint) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/DashboardServerFixture.cs:61]: that returns a ValueTask, I changed to return `DashboardApp.DisposeAsync().AsTask()`

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/AppBarTests.cs
Comment (mxschmitt) [inline tests/Aspire.Dashboard.Tests/Integration/Playwright/AppBarTests.cs:40]: Remove the delay and use our assertion which retries? https://playwright.dev/dotnet/docs/api/class-locatorassertions#locator-assertions-to-have-attribute

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs
Comment (mxschmitt) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs]: Usually we strongly recommend doing install and --with-deps from the outside. Since these require sudo and modify the system. These should not delay the test execution and skipping it on Linux seems like showing the drawbacks for it.

lets call it from the CI yaml instead?

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs
Comment (mxschmitt) [inline tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:45]: I recommend page -> Page like we have in our base classes.

```suggestion
        await Page.GotoAsync("/");
```

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs
Comment (mxschmitt) [inline tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:47]: Is this like a custom retry strategy? I recommend to e.g. stick with ToBeVisible which will retry and end up in better reporting in e.g. Trace Viewer or logs.

https://playwright.dev/dotnet/docs/api/class-locatorassertions#locator-assertions-to-be-visible

**Comment by adamint** [comment]
> Usually we strongly recommend doing install and --with-deps from the outside. Since these require sudo and modify the system. These should not delay the test execution and skipping it on Linux seems like showing the drawbacks for it.

> lets call it from the CI yaml instead?

This code was removed.

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs
Comment (adamint) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:45]: As a method parameter, that does not follow naming conventions. If `page` were a property, I would agree

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs
Comment (adamint) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs]: Thanks, I've made this change 😄

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs
Comment (mxschmitt) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:26]: I see, thought its a property on the class where page gets created in TestInitialize etc. Feel free to ignore then.

**Comment by danmoseley** [comment]
Thanks @adamint . Does any of my feedback from your original PR apply here? (IDK)

File: tests/Aspire.Dashboard.Tests/Aspire.Dashboard.Tests.csproj
Comment (radical) [inline tests/Aspire.Dashboard.Tests/Aspire.Dashboard.Tests.csproj:18]: If this was added for getting `PlaywrightProvider`, then instead use `<Import Project="..\Shared\Playwright\Playwright.targets" />`

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/AppBarTests.cs
Comment (radical) [inline tests/Aspire.Dashboard.Tests/Integration/Playwright/AppBarTests.cs:23]: nit: this can return `RunTestAsync(..)` directly.

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightTestsBase.cs
Comment (radical) [inline tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightTestsBase.cs:37]: This will be using a single context for all the tests in a class. Is that intentional?

Comment (radical) [reply tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightTestsBase.cs:37]: nvm, that makes sense at least for a single class at least.

File: tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs
Comment (radical) [inline tests/Aspire.Dashboard.Tests/Integration/Playwright/PlaywrightFixture.cs:17]: This would create the browser instance, and launch the browser for every test class using this fixture. For workload tests we use:
https://github.com/adamint/aspire/blob/64ea4ab15976fbf903c9663aa6982205be8b1e2f/tests/Aspire.Workload.Tests/WorkloadTestsBase.cs#L22-L34
and `private static Lazy<IBrowser> Browser => new(CreateBrowser);`

**Comment by radical** [comment]
Screenshot 2024-06-28 at 20 43 46

And the test was skipped was linux.

**Comment by radical** [comment]
@adamint I merged the base playwright changes in a separate PR, and update this one to fix the merge conflicts.

**Comment by adamint** [comment]
> @adamint I merged the base playwright changes in a separate PR, and update this one to fix the merge conflicts.

Is anything else needed here? @radical 
thank you 😄 

@danmoseley not that I'm aware of. The recommendation is to avoid assuming the structure of the page, so locating elements by text instead of css classes is preferred

**Review by radical — APPROVED** [review]
The infra bits LGTM 👍

**Pull request merged**
